core/types/src/keys.rs

```diff
@@ -0,0 +1,356 @@
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        let data = self.key.to_bytes();
+        for d in data {
+            write!(f, "{:02x}", d)?;
+        }
+        Ok(())
+    }
+}
+
+// Shared private key methods
+
+impl<ADDR, KIND> Key<ADDR, KIND, RistrettoPrivate> {
+    /// Fetch private key bytes
+    pub fn to_bytes(&self) -> [u8; 32] {
+        self.key.to_bytes()
+    }
+}
+
+/// Create a private key from [`RistrettoPrivate`] object
+impl<ADDR, KIND> From<RistrettoPrivate> for Key<ADDR, KIND, RistrettoPrivate> {
+    fn from(p: RistrettoPrivate) -> Self {
+        Self {
+            key: p,
+            _addr: PhantomData,
+            _kind: PhantomData,
+        }
+    }
+}
+
+/// Fetch corresponding public key for a private key object
+impl<ADDR, KIND> From<Key<ADDR, KIND, RistrettoPrivate>> for Key<ADDR, KIND, RistrettoPublic> {
+    fn from(p: Key<ADDR, KIND, RistrettoPrivate>) -> Self {
+        Self {
+            key: RistrettoPublic::from(&p.key),
+            _addr: PhantomData,
+            _kind: PhantomData,
+        }
+    }
+}
+
+/// Attempt to create a private key from a compressed point, wrapping
+/// [`RistrettoPrivate::try_from`]
+impl<ADDR, KIND> TryFrom<&[u8; 32]> for Key<ADDR, KIND, RistrettoPrivate> {
+    type Error = KeyError;
+
+    fn try_from(p: &[u8; 32]) -> Result<Self, Self::Error> {
+        let key = RistrettoPrivate::try_from(p)?;
+        Ok(Self {
+            key,
+            _addr: PhantomData,
+            _kind: PhantomData,
+        })
+    }
+}
+
+/// Access underlying [`Scalar`] for private keys using [`AsRef`]
+impl<ADDR, KIND> AsRef<Scalar> for Key<ADDR, KIND, RistrettoPrivate> {
+    fn as_ref(&self) -> &Scalar {
+        self.key.as_ref()
+    }
+}
+
+/// Create private keys from raw [`Scalar`]
+impl<ADDR, KIND> From<Scalar> for Key<ADDR, KIND, RistrettoPrivate> {
+    fn from(s: Scalar) -> Self {
+        Self {
+            key: RistrettoPrivate::from(s),
+            _addr: PhantomData,
+            _kind: PhantomData,
+        }
+    }
+}
+
+/// [`PartialEq`] via public key conversion for Private key objects
+impl<ADDR, KIND> PartialEq for Key<ADDR, KIND, RistrettoPrivate> {
+    fn eq(&self, other: &Self) -> bool {
+        RistrettoPublic::from(&self.key) == RistrettoPublic::from(&other.key)
+    }
+}
+
+/// PartialEq for backwards compatibility with private key objects
+impl<ADDR, KIND> PartialEq<RistrettoPrivate> for Key<ADDR, KIND, RistrettoPrivate> {
+    fn eq(&self, other: &RistrettoPrivate) -> bool {
+        RistrettoPublic::from(&self.key) == RistrettoPublic::from(other)
+    }
+}
+
+/// PartialEq for backwards compatibility with private key objects
+impl<ADDR, KIND> PartialEq<Key<ADDR, KIND, RistrettoPrivate>> for RistrettoPrivate {
+    fn eq(&self, other: &Key<ADDR, KIND, RistrettoPrivate>) -> bool {
+        RistrettoPublic::from(self) == RistrettoPublic::from(&other.key)
+    }
+}
+
+/// [`core::fmt::Display`] for public key objects
```

Comment:
:upside_down_face: 
Maybe the intent was to document the output is converted to a public key.
```suggestion
/// [`core::fmt::Display`] for private key objects
```